Framework/CurveFitting/src/Functions/CrystalFieldFunction.cpp

```diff
@@ -0,0 +1,1320 @@
+  m_nControlSourceParams = m_nControlParams + m_source->nParams();
+}
+
+/// Update spectrum function if necessary.
+void CrystalFieldFunction::checkTargetFunction() const {
+  if (m_dirtyTarget) {
+    updateTargetFunction();
+  }
+  if (!m_target) {
+    throw std::logic_error(
+        "CrystalFieldFunction failed to generate target function.");
+  }
+}
+
+/// Uses source to calculate peak centres and intensities
+/// then populates m_spectrum with peaks of type given in PeakShape attribute.
+void CrystalFieldFunction::buildTargetFunction() const {
+  checkSourceFunction();
+  m_dirtyTarget = false;
+  if (isMultiSite()) {
+    buildMultiSite();
+  } else {
+    buildSingleSite();
+  }
+  m_attributeNames.clear();
+}
+
+/// Build the target function in a single site case.
+void CrystalFieldFunction::buildSingleSite() const {
+  if (isMultiSpectrum()) {
+    buildSingleSiteMultiSpectrum();
+  } else {
+    buildSingleSiteSingleSpectrum();
+  }
+}
+
+/// Build the target function in a multi site case.
+void CrystalFieldFunction::buildMultiSite() const {
+  if (isMultiSpectrum()) {
+    buildMultiSiteMultiSpectrum();
+  } else {
+    buildMultiSiteSingleSpectrum();
+  }
+}
+
+/// Build the target function in a single site - single spectrum case.
+void CrystalFieldFunction::buildSingleSiteSingleSpectrum() const {
+  auto spectrum = new CompositeFunction;
+  m_target.reset(spectrum);
+  m_target->setAttributeValue("NumDeriv", true);
+  auto bkgdShape = getAttribute("Background").asUnquotedString();
+  bool fixAllPeaks = getAttribute("FixAllPeaks").asBool();
+
+  if (!bkgdShape.empty()) {
+    auto background =
+        API::FunctionFactory::Instance().createInitialized(bkgdShape);
+    spectrum->addFunction(background);
+  }
+
+  FunctionDomainGeneral domain;
+  FunctionValues values;
+  m_source->function(domain, values);
+
+  if (values.size() == 0) {
+    return;
+  }
+
+  if (values.size() % 2 != 0) {
+    throw std::runtime_error(
+        "CrystalFieldPeaks returned odd number of values.");
+  }
+
+  auto xVec = m_control.getAttribute("FWHMX").asVector();
+  auto yVec = m_control.getAttribute("FWHMY").asVector();
+  auto &FWHMs = m_control.FWHMs();
+  auto defaultFWHM = FWHMs.empty() ? 0.0 : FWHMs[0];
+
+  auto fwhmVariation = getAttribute("FWHMVariation").asDouble();
+  auto peakShape = getAttribute("PeakShape").asString();
+  size_t nRequiredPeaks = getAttribute("NPeaks").asInt();
+  CrystalFieldUtils::buildSpectrumFunction(*spectrum, peakShape, values, xVec,
+                                           yVec, fwhmVariation, defaultFWHM,
+                                           nRequiredPeaks, fixAllPeaks);
+}
+
+/// Build the target function in a single site - multi spectrum case.
+void CrystalFieldFunction::buildSingleSiteMultiSpectrum() const {
+  auto fun = new MultiDomainFunction;
+  m_target.reset(fun);
+
+  DoubleFortranVector en;
+  ComplexFortranMatrix wf;
+  ComplexFortranMatrix ham;
+  ComplexFortranMatrix hz;
```

Comment:
what is hz?

Reply:
It's the Zeeman Hamiltonian term (the magnetic field dependent part) - `ham` is the crystal field Hamiltonian in zero external field. I guess it could be changed to `hzeeman` or something like that...

Reply:
I think thats a good idea. I thought it may have been the z component of the Hamiltonian.